Add an `include` keyword (#174)
WebAssembly/component-model: design/mvp/WIT.md

    export c: self.c
}

// union.wit
world union-my-world-1 {
    include pkg.my-world-1 with { a as a1, b as b1 }
    include pkg.my-world-2
}

world union-my-world-2 {
    // resolve conflicts
    import a1: pkg.my-world-1.a1
    import b1: pkg.my-world-1.b1
    export d: pkg.my-world-1.d

    import a: pkg.my-world-2.a2
    import b: pkg.my-world-2.b2
    export c: pkg.my-world-2.c
}
```

### De-duplication

If two interfaces have the same structure, then these two interfaces are considered to be structurally equivalent. The `include` statement can
deduplicate the imports and exports of the included worlds if the they are structurally equivalent following the [Subtyping](Subtyping.md) rules. For example, the following world `union-my-world-3` is equivalent to `union-my-world-4`:

```wit
// my-world-1.wit
world my-world-1 {
    import a1: pkg.a
    import b1: pkg.b
    export c1: pkg.c
}

// my-world-2.wit
world my-world-2 {
    import a2: pkg.a
    import b2: pkg.b
    export c2: pkg.c
}

// union.wit
world union-my-world-3 {
    include pkg.my-world-1
    include pkg.my-world-2
}

world union-my-world-4 {
    import a1: pkg.a
    import b1: pkg.b
    export c1: pkg.c

> }

Reviewer: This I think runs a risk of being surprising because `my-world-2` is effectively ignored here. Could this perhaps start out with a rule that if two interfaces are imported they must be imported under the same name? That means that the union-world above would be an error by default, but it could be fixed with `with` to do renaming to make sure the names line up.

Author: > if two interfaces are imported they must be imported under the same name?

that seems reasonable to me.

Author: I have updated the doc to address your comment.

### De-duplication with `with`

When two worlds have both name conflicts and structurally equivalent imports and exports, the semantics of `include` will do deduplication first and then resolve the name conflicts with `with` statements. For example, the following world `union-my-world-5` is equivalent to `union-my-world-6`:

```wit

// my-world-1.wit
world my-world-1 {
    import a1: pkg.a
    import b1: pkg.b
    export c: pkg.c
}

// my-world-2.wit
world my-world-2 {
    import a1: pkg.a
    import b1: pkg.b
    export c: pkg.d
}

// union
world union-my-world-5 {
    include pkg.my-world-1 with { c as c1 }
    include pkg.my-world-2
}

world union-my-world-6 {
    import a1: pkg.a
    import b1: pkg.b
    export c1: pkg.c
    export c: pkg.d
}

```

### A Note on SubTyping

As of now, host bindings are required to implicitly interpret all exports as optional to make a component targeting an included World a subtype of the union World. This [comment](https://github.com/WebAssembly/component-model/issues/169#issuecomment-1446776193) describes the reasoning behind this decision.

> In the future, when `optional` export is supported, the world author may explicitly mark exports as optional to make a component targeting an included World a subtype of the union World.

Author: Oh I had a different interpretation of Luke's comment. I thought the goal is that we want to produce subtyped worlds and a solution to that is described by Luke as "In the short term, I expect we could get away with simply defining include to take the union of exports (just like imports) and having host bindings implicitly interpret all exports as optional"

@lukewagner please correct me if this is wrong.

> In the future, when `optional` export is supported, the world author may explicitly mark exports as optional to make a component targeting an included World a subtype of the union World.

Reviewer: Currently there's no support for optional exports, nor is there any planned in the near future at this time. If the goal of `include` is to produce subtyped worlds then this will need to be put on hold until optional exports are considered.

> In the future, when `optional` export is supported, the world author may explicitly mark exports as optional to make a component targeting an included World a subtype of the union World.

Author: @alexcrichton ACK, I am okay with proceeding without subtyped worlds

> In the future, when `optional` export is supported, the world author may explicitly mark exports as optional to make a component targeting an included World a subtype of the union World.

Author: I have updated the doc to address your comment.

Below is a simple example of a world that includes two other worlds.

```wit
// worlds.wit
world my-world-1 {
    import a: self.a
    import b: self.b
    export c: self.c
}

world my-world-2 {
    import foo: self.foo
    import bar: self.bar
    export baz: self.baz
}

world union-my-world {
     include self.my-world-1
     include self.my-world-2
}
```

The `include` statement is used to include the imports and exports of another World to the current World. It says that the new World should be able to run all components that target the included worlds and more.

The `union-my-world` World defined above is equivalent to the following World:

```wit
world union-my-world {
    import a: self.a
    import b: self.b
    export c: self.c
    import foo: self.foo
    import bar: self.bar
    export baz: self.baz
}
```

The `include` statement also works with [WIT package](#wit-packages-and-use) defined below with the same semantics. For example, the following World `union-my-world-1` is equivalent to `union-my-world-2`:

```wit
// b.wit
interface b { ... }

// a.wit

> interface a { ... }

Author: Fixed it. Thank you!

world my-world-1 {
    import a: self.a 
    import b: pkg.b
    import c: io.c // external package
    export d: interface exp { ... }
}

// union.wit

world union-my-world-1 {

>     include pkg.my-world-1

Author: Yes that's right. Thanks!

```wit
// b.wit
interface b { ... }

// a.wit
interface a {}

world my-world-a {
    import a
    import b
    import wasi:io/c
    export d: interface { ... }
}

// union.wit
package local:demo

world union-my-world-a {
    include my-world-a
}

world union-my-world-b {
    import a
    import b
    import wasi:io/c

    export d: interface { ... }
}
```

### Name Conflicts

When two or more included Worlds have the same name for an import or export, the name is considered to be in conflict. The conflict needs to be explicitly resolved by the world author using the `with` keyword. `with` allows the world author to rename the import or export to a different name.

The following example shows how to resolve name conflicts where `union-my-world-a` and `union-my-world-b` are equivalent:

```wit
package local:demo

world world-one { import a: func() }
world world-two { import a: func() }

world union-my-world-a { 
    include foo
    include bar with { a as b }
}

world union-my-world-b {
  import a: func()
  import b: func()
}
```

> ### De-duplication

Reviewer 2: Perhaps this should go before the name conflicts section? It would also be worth clarifying how deduplication and name conflicts can be both supported. For example, can `with` syntax be used in cases where deduplication would otherwise have occured? I'd assume so but it would be good to clarify this.

Author: I should clarify. Deduplication only applies to IDs, and `with` syntax only applies to kebab-names.

    import a
    import b
    export c
    import foo
    import bar
    export baz
}
```

The `include` statement also works with [WIT package](#wit-packages-and-use) defined below with the same semantics. For example, the following World `union-my-world-a` is equivalent to `union-my-world-b`:

```wit
package local:demo

interface b { ... }
interface a { ... }

world my-world-a {
    import a
    import b
    import wasi:io/c
    export d: interface { ... }
}

world union-my-world-a {
    include my-world-a
}

world union-my-world-b {
    import a
    import b
    import wasi:io/c

    export d: interface { ... }
}
```

### De-duplication of IDs

If two worlds shared the same set of import and export IDs, then the union of the two worlds will only contain one copy of this set. For example, the following two worlds `union-my-world-a` and `union-my-world-b` are equivalent:

```wit
package local:demo

world my-world-a {
    import a1
    import b1
}

world my-world-b {
    import a1
    import b1
}

world union-my-world-a {
    include my-world-a
    include my-world-b
}

world union-my-world-b {
    import a1
    import b1
}
```

### Name Conflicts and `with`

> When two or more included Worlds have the same name for an import or export, the name is considered to be in conflict. The conflict needs to be explicitly resolved by the world author using the `with` keyword. `with` allows the world author to rename the import or export to a different name.
> 
> Notice that when import or export names are IDs and since IDs are unique, there is no need to resolve name conflicts. Thus the `with` syntax is invalid when used with `include <ID>`. Only when import or export names are kebab names, name conflicts need to be resolved.
> 

Reviewer 3: ```suggestion
When two or more included Worlds have the same name for an import or export that does *not* have an ID, de-duplication cannot be used (because the two same-named imports/exports might have different meanings in the different worlds) and thus the conflict has to be resolved using the `with` keyword:
```

The following example shows how to resolve name conflicts where `union-my-world-a` and `union-my-world-b` are equivalent:

```wit
package local:demo

world world-one { import a: func() }
world world-two { import a: func() }

world union-my-world-a { 
    include foo
    include bar with { a as b }
}

world union-my-world-b {
  import a: func()
  import b: func()
}
```

> The following example shows an invalid example that `with` is used when the import or export name is an ID:
> 
> ```wi
> package local:demo
> 
> world my-world-a {
>     import a1
>     import b1
> }
> 
> world my-world-b {
>     import a1
>     import b1
> }
> 
> world union-my-world-a {
>     include my-world-a with { a1 as a3 }
>     include my-world-b with { a1 as a2 }
> }
> ```

Reviewer 3: ```suggestion
`with` cannot be used to rename IDs, however, so the following world would be invalid:
```wit
package local:demo

interface a {
  foo: func()
}

world world-using-a {
    import a
}

world invalid-union-world {
    include my-using-a with { a as b }  // invalid: 'a', which is short for 'local:demo/a', is an ID
}
```
```

A World can be created by taking the union of two or more worlds. This operation allows world builders to form larger worlds from smaller worlds.

Below is a simple example of a world that includes two other worlds.

```wit
package local:demo

// definitions of a, b, c, foo, bar, baz are omitted

world my-world-a {
    import a
    import b
    export c
}

world my-world-b {
    import foo
    import bar
    export baz
}

world union-my-world {
     include my-world-a
     include my-world-b
}
```

The `include` statement is used to include the imports and exports of another World to the current World. It says that the new World should be able to run all components that target the included worlds and more.

The `union-my-world` World defined above is equivalent to the following World:

```wit
world union-my-world {
    import a
    import b
    export c
    import foo
    import bar
    export baz
}
```

The `include` statement also works with [WIT package](#wit-packages-and-use) defined below with the same semantics. For example, the following World `union-my-world-a` is equivalent to `union-my-world-b`:

```wit
// b.wit
interface b { ... }

// a.wit

> interface a {}

Reviewer 3: Also, a suggestion: with the new Wit syntax, the filenames no longer matter, so it's just as well to remove the "// a.wit" comments since they're not relevant

> interface a {}

Author: Done

world my-world-a {
    import a1
    import b1
}

world my-world-b {
    import a1
    import b1
}

world union-my-world-a {
    include my-world-a
    include my-world-b
}

world union-my-world-b {
    import a1
    import b1
}
```

### Name Conflicts and `with`

When two or more included Worlds have the same name for an import or export, the name is considered to be in conflict. The conflict needs to be explicitly resolved by the world author using the `with` keyword. `with` allows the world author to rename the import or export to a different name.

Notice that when import or export names are IDs and since IDs are unique, there is no need to resolve name conflicts. Thus the `with` syntax is a no-op in this case. Only when import or export names are kebab names, name conflicts need to be resolved.

The following example shows how to resolve name conflicts where `union-my-world-a` and `union-my-world-b` are equivalent:

```wit
package local:demo

world world-one { import a: func() }
world world-two { import a: func() }

world union-my-world-a { 
    include foo
    include bar with { a as b }
}

world union-my-world-b {
  import a: func()
  import b: func()
}
```

> The following example shows that `with` is a no-op when the import or export name is an ID:

Reviewer 2: Rather than a no-op can we just make this throw - that is, it is an error to refer to ids at all in this process? Also helps ensure any future compat concerns.

Reviewer 3: +1

Author: Okay, made the `include <ID> with { ... }` case invalid.

Author: I am happy to send a PR to wasm-tools to adopt this decision.

Reviewer 3: Thanks!

The `interface` item here additionally defines the grammar for IDs used to refer
to `interface` items.

## Item: `include`

> A `include` statement enables union the current world with another world. The structure of an `include` statement is:

Reviewer 2: ```suggestion
A `include` statement enables the union of the current world with another world. The structure of an `include` statement is:
```

```wit
include pkg.my-world-1 with { a as a1, b as b1 }

> include self.my-world-2

Reviewer 2: Note these are still using the dated syntax.

Author: Ah missed it. Thanks!

The `interface` item here additionally defines the grammar for IDs used to refer
to `interface` items.

## Item: `include`

A `include` statement enables the union of the current world with another world. The structure of an `include` statement is:

```wit
include wasi:io/my-world-1 with { a as a1, b as b1 }
include my-world-2
```

```ebnf
include-item ::= 'include' use-path
               | 'include' use-path 'with' '{' include-names-list '}'

include-names-list ::= include-names-item

>                  | include-names-item ',' include-names-item?

Reviewer 3: ```suggestion
                     | include-names-item ',' include-names-item?
```
